### ISSUE TITLE
Spyware claim

### ISSUE THREAD

**Issue by EliezerBee** [skechboy/SpeakIt#12]
On [this Chrome extension's page](https://chrome.google.com/webstore/detail/speakit/pgeolalilifpodheeocdmbhehgnkkbak/reviews) someone named Jason Ryan wrote this:

> _This extension now includes a spyware. It sends all my visited links and browsing habits to a remote server via upalytics javascript code. I did not receive a notice or warning from the extension that my traffic would be monitored before installation. Google should take this spyware down ASAP._

Could the developer please comment on this?
Thank you!

**Comment by JonasNo** [comment]
Note: I'm not connected to this project.
There is a file called "upalytics.js" from "Upalytics.com - Analytics JS for Chrome" in the extension and is active no matter if the option "_Automatically send anonymous usage data, to improve SpeakIt! functionality_" is turned on or off.
There have been shady analytics scripts active before too. Without any information being given to the users.

The script sends your navigation history. Where you navigate to and where you were before.

At the start of the extension (browser start or extension is enabled) it sends some for of ID, install date and version of the extension twice. ID only different.

```
"https://fu-61495ceefa.upalytics.com/dynamic?s=940&ins=1234567891.601&ver=0.2.988"
and
"https://fu-61495ceefa.upalytics.com/dynamic?s=1940&ins=1234567891.601&ver=0.2.988"

```

When you navigate it sends each navigation.

Example:
When navigating from "https://www.youtube.com/" to "http://translate.google.com/"

It sends information like this but base64 encoded. The "X"s is me censoring out possibly identifiable data.

```
"s=1XX0&md=XX&pid=XXXXXXXXXXXXXXX&sess=XXXXXXXXXXXXXXXXX0&sub=chrome&q=https%3A//translate.google.com/&prev=https%3A//www.youtube.com/&tmv=XXX7&tmf=1&sr=http%3A//translate.google.com/"
to
"https://fu-61495ceefa.upalytics.com/service2"

```
and

```
"s=XX0&md=X1&pid=XXXXXXXXXXXXXXX&sess=XXXXXXXXXXXXXXXXXX&q=https%3A%2F%2Ftranslate.google.com%2F&prev=https%3A%2F%2Fwww.youtube.com%2F&link=0&sub=chrome&hreferer=&tmv=XX17"
to
"https://fu-61495ceefa.upalytics.com/related"

```

Basically it's Windows 10 but without sending the text it speaks. Curiously enough this "anonymous" usage data can not be used to improve the extension. It's not sending any real usage data.

I noticed this script and manually edited the extension locally and removed upalytics. Works fine without it.
Personally I check each update for shady things. Just in case. I do that with all the extensions I use.

Hope this information helps you.
Be paranoid. Verify everything. Stay safe.

**Comment by EliezerBee** [comment]
JonasNo, thank you for your very helpful answer. Is there a reasonable straightforward way to explain to a non-extension developer (I'm a C# developer) how to edit the extension locally?

**Comment by skechboy** [comment]
Hello,

thanks for your interest in SpeakIt! you're right in the latest update I've added analytic code from http://upalytics.com/ As explained from Jonas the code is sending random and anonymous usage data to upalytics server. 

I'm using this data for leveraging user habits and for making this extension works better for most visited sites, however I'm aware that someone can see this as threat to their privacy therefor in the next version all users will be asked if they want to join this. 

In meanwhile if you remove the code manualy it won't affect extensions behavior in any way.

Thanks

**Comment by EliezerBee** [comment]
Skechboy, thank you for your great extension!
When you wrote

> in the next version all users will be asked if they want to join this

did you mean:

1. there will be an option to use the extension with the Upalytics part disabled, or
2. there will be a warning saying that by using the extension you understand and agree that it's using Upalytics - but there will be no choice whether to enable or disable that functionality
?

**Comment by skechboy** [comment]
My idea is to show a screen where the user will be asked if they want to send anonymous usage data.

So basically they will have Yes and No, by default will be selected No so the user will actually need to Opt in to send anonymous data usage, I think this is the best way to achieve this.

If you have any other suggestions or ideas regards the tracking let me know.

Thanks

**Comment by EliezerBee** [comment]
That sounds great! Any idea when this version might come out?

**Comment by skechboy** [comment]
I'm preparing it for this weekend i'm sure that it will be out in 2-3 days.

**Comment by EliezerBee** [comment]
That's great!
Any chance you could also address the "[apostrophe issue](https://github.com/skechboy/SpeakIt/issues/7)" at the same time?
